### PR TITLE
Compute sizedness with a fixed-point analysis

### LINKED ISSUE #768: Rewrite `is_unsized` as a fix-point analysis
See https://github.com/servo/rust-bindgen/issues/536 for details.

This is very similar to when we re-wrote `has_destructor` to a fixed-point analysis, and that PR could be used to guide fixing this issue: https://github.com/rust-lang-nursery/rust-bindgen/pull/850

### PR DESCRIPTION
**Pull request by fitzgen**
This fixes a couple bugs where we weren't properly adding an `_address` byte. It also helps pave the way for computing implicit fields (such as padding, `_address`, vtable pointers, etc) in its own pass, before codegen.

Fixes #768

r? @pepyakin

### CODE REVIEW

File: src/ir/analysis/sizedness.rs [hunk 1]
```diff
@@ -0,0 +1,360 @@
...
+            TypeKind::BlockPointer |
+            TypeKind::ObjCId |
+            TypeKind::ObjCSel |
+            TypeKind::Pointer(..) => {
+                trace!("    {:?} is known not to be zero-sized", ty.kind());
+                self.insert(id, SizednessResult::NonZeroSized)
+            }
+
+            TypeKind::ObjCInterface(..) => {
+                trace!("    I guess obj-c interfaces are not zero-sized?");
```
Comment (pepyakin) [inline src/ir/analysis/sizedness.rs]: FWIW, as far as I know, every obj-c "interface" have an `isa` pointer.

File: src/ir/analysis/sizedness.rs [hunk 2]
```diff
@@ -0,0 +1,360 @@
...
+            TypeKind::Array(..) => {
+                trace!("    arrays of > 0 elements are not zero-sized");
+                self.insert(id, SizednessResult::NonZeroSized)
+            }
+
+            TypeKind::Comp(ref info) => {
+                trace!("    comp considers its own fields and bases");
+
+                if !info.fields().is_empty() {
+                    return self.insert(id, SizednessResult::NonZeroSized);
```
Comment (pepyakin) [inline src/ir/analysis/sizedness.rs]: But what if all fields are zero sized 🤔 ?

Comment (fitzgen) [reply src/ir/analysis/sizedness.rs]: They can't be, because of the every-object-must-be-addressable rules. Only base members can truly be zero sized, because its the "same" object.

File: src/ir/analysis/sizedness.rs (hunk 1 above)
Comment (fitzgen) [reply src/ir/analysis/sizedness.rs]: Cool, I'll note this. FWIW, I was just copying existing behavior.

**Comment by fitzgen** [comment]
Thanks for the review @pepyakin !

File: src/ir/analysis/sizedness.rs (hunk 2 above)
Comment (fitzgen) [reply src/ir/analysis/sizedness.rs]: That is, with

```c++
struct Empty {};

struct Inherits : public Empty {
    int x;
};

struct Contains {
    Empty empty;
    int x;
};
```

`sizeof(Contains)` is one greater than `sizeof(Inherits)`

Comment (pepyakin) [reply src/ir/analysis/sizedness.rs]: yeah, makes sense. Thank you!

**Comment by fitzgen** [comment]
I added that example, and another for arrays of zero-sized types, as tests.

Thanks again for review @pepyakin !

@bors-servo r=pepyakin

**Comment by fitzgen** [comment]
Added one more test for arrays of zero elements and flexible arrays also.

@bors-servo r=pepyakin
